### PR TITLE
feat(browser): isolate browser tests

### PR DESCRIPTION
**Pull request by userquin**
This PR isolates tests creating an iframe per test file.

On my local tests, I need to start the tests a few times to get it working, I have added vue and react examples and vue should also work (this is one we need to revert changes).

DRAFT: we'll need to revert some changes in existing tests and cleanup some modules.

### CODE REVIEW

**Comment by sheremet-va** [comment]
> > creating an iframe per test file
> 
> There can be only 1 focused element per page, so I don't think iframes are good enough for isolation. Maybe 1 page per test? Not sure if it's possible to preserve/emulate focus like with the devtools' "Emulate a focused page" command.

This is a valid concern. We need to add a test with focusing elements.

**Comment by userquin** [comment]
> > > creating an iframe per test file
> > 
> > 
> > There can be only 1 focused element per page, so I don't think iframes are good enough for isolation. Maybe 1 page per test? Not sure if it's possible to preserve/emulate focus like with the devtools' "Emulate a focused page" command.
> 
> This is a valid concern. We need to add a test with focusing elements.

I'll try to add a focus test...

**Comment by userquin** [comment]
working

imagen

**Comment by userquin** [comment]
To run browser test:
1) from root `ni && nr build` and then
2) open a new terminal, change to `examples/vue-browser`  and run `nr test`
3) open a new terminal, change to `examples/react-testing-lib-browser`  and run `nr test`

I'll try to add more frameworks examples: lit, svelte, preact and solid, and maybe also sveltekit

File: packages/browser/src/client/tester.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    config.snapshotOptions.snapshotEnvironment = new BrowserSnapshotEnvironment()
+
+  try {
+    await setupCommonEnv(config)
+
+    await startTests([filename], runner)
+  }
+  finally {
+    // TODO: remove version once logs removed from main.ts module
+    channel.postMessage({ type: 'done', filename, version })
```
Comment (userquin) [inline packages/browser/src/client/tester.ts:46]: Remove version here

File: packages/vitest/src/node/pools/browser.ts [hunk 2]
```diff
@@ -41,8 +44,9 @@ export function createBrowserPool(ctx: Vitest): ProcessPool {
     const origin = `http://${ctx.config.browser.api?.host || 'localhost'}:${project.browser.config.server.port}`
     const paths = files.map(file => relative(project.config.root, file))
 
-    const isolate = project.config.isolate
-    if (isolate) {
+    // TODO@browser: review this
+    // const isolate = project.config.isolate
+    /* if (isolate) {
```
Comment (userquin) [inline packages/vitest/src/node/pools/browser.ts:49]: Remove old code

File: packages/browser/src/client/main.ts [hunk 3]
```diff
@@ -1,141 +1,204 @@
...
+      const idx = waitingPaths.indexOf(filename)
+      if (idx === -1)
+        return
+
+      waitingPaths.splice(idx, 1)
+
+      if (!waitingPaths.length) {
+        await rpcDone()
+        await rpc().onDone('no-isolate')
+        listenToRun = true
```
Comment (userquin) [inline packages/browser/src/client/main.ts:187]: Use try/finally?

File: packages/vitest/src/types/browser.ts [hunk 4]
```diff
@@ -53,6 +54,13 @@ export interface BrowserConfigOptions {
    */
   api?: ApiConfig | number
 
+  /**
+   * When running on browser, enable this flag to show the result.
+   *
+   * @default false
+   */
+  enableUI?: boolean
```
Comment (Aslemammad) [inline packages/vitest/src/types/browser.ts:62]: @userquin  Do we actually need this? I don't have any strong preference over having or not having it as an option, just asking. Can we show it without an option?

Comment (userquin) [reply packages/vitest/src/types/browser.ts:62]: You can run test in the ui and those test may not have ui, check test/browser

Comment (userquin) [reply packages/vitest/src/types/browser.ts:62]: It is read in the ui when loading the configuration, it is used to show the tab in the details panel

Comment (userquin) [reply packages/vitest/src/types/browser.ts:62]: @Aslemammad 

imagen

Comment (Aslemammad) [reply packages/vitest/src/types/browser.ts:62]: I see, That's better, thanks.

**Comment by sheremet-va** [comment]
@userquin can you resolve conflicts last time?

File: packages/vitest/src/types/browser.ts (hunk 4 above)
Comment (sheremet-va) [reply packages/vitest/src/types/browser.ts:64]: I think the use case when you run it in the browser without a UI is not so popular. `@vitest/browser` is primarily to allow component testing. examples/browser is just a test to make sure everything works seamlessly

**Review by sheremet-va — CHANGES_REQUESTED** [review]

**Comment by userquin** [comment]
> @userquin can you resolve conflicts last time?

Sure

**Comment by jfgreffier** [comment]
Closes #3154

**Comment by shairez** [comment]
Looks like an awesome feature!
Is there anything blocking it other than fixing the conflicts?

**Comment by DanWebb** [comment]
Is the browser UI tab still a planned addition after #5036? 🤞🙏

**Pull request closed without merging**